pyop2/caching.py

```diff
@@ -440,119 +441,89 @@ class DEFAULT_CACHE(dict):
-                # Create a PyOP2 comm associated with the key, so it is decrefed when the wrapper exits
-                with temp_internal_comm(comm_fetcher(*args, **kwargs)) as comm:
-                    # Fetch the per-comm cache_collection or set it up if not present
-                    # A collection is required since different types of cache can be set up on the same comm
-                    cache_collection = comm.Get_attr(comm_cache_keyval)
-                    if cache_collection is None:
-                        cache_collection = {}
-                        comm.Set_attr(comm_cache_keyval, cache_collection)
-                    # If this kind of cache is already present on the
-                    # cache_collection, get it, otherwise create it
-                    local_cache = cache_collection.setdefault(
-                        (cf := cache_factory()).__class__.__name__,
-                        cf
-                    )
-                    local_cache = cache_collection[cf.__class__.__name__]
-
-                    # If this is a new cache or function add it to the list of known caches
-                    if (comm, comm.name, func, local_cache) not in [(c.comm, c.comm_name, c.func, c.cache()) for c in _KNOWN_CACHES]:
-                        # When a comm is freed we do not hold a reference to the cache.
-                        # We attach a finalizer that extracts the stats before the cache
-                        # is deleted.
-                        _KNOWN_CACHES.append(_CacheRecord(next(_CACHE_CIDX), comm, func, local_cache))
-
-                    # Grab value from all ranks cache and broadcast cache hit/miss
-                    value = local_cache.get(key, CACHE_MISS)
-                    debug_string = f"{COMM_WORLD.name} R{COMM_WORLD.rank}, {comm.name} R{comm.rank}: "
-                    debug_string += f"key={k} in cache: {local_cache.__class__.__name__} cache "
-                    if value is CACHE_MISS:
-                        debug(debug_string + "miss")
-                        cache_hit = False
-                    else:
-                        debug(debug_string + "hit")
-                        cache_hit = True
-                    all_present = comm.allgather(cache_hit)
-
-                    # If not present in the cache of all ranks we force re-evaluation on all ranks
-                    if not min(all_present):
-                        value = CACHE_MISS
+def parallel_cache(
+    hashkey=default_parallel_hashkey,
+    comm_getter=default_comm_getter,
+    cache_factory=lambda: DEFAULT_CACHE(),
+    bcast=False,
+):
+    """Parallel cache decorator.
+
+    Parameters
+    ----------
+    hashkey :
+        Callable taking ``*args`` and ``**kwargs`` and returning a hash.
+    comm_getter :
+        Callable taking ``*args`` and ``**kwargs`` and returning the
+        appropriate communicator.
+    cache_factory :
+        Callable that will build a new cache (if one does not exist).
```

Comment:
```suggestion
    cache_factory :
        Callable that will build a new cache (if one does not exist).
        This will be called every time the decorated function is called, and must return an instance
        of the same type every time it is called.
```